src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionService.cs

```diff
@@ -0,0 +1,517 @@
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion.Log;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal sealed class SerializableImportCompletionItem
+    {
+        public readonly string SymbolKeyData;
+        public readonly int Arity;
+        public readonly string Name;
+        public readonly Glyph Glyph;
+        public readonly string ContainingNamespace;
+
+        public SerializableImportCompletionItem(string symbolKeyData, string name, int arity, Glyph glyph, string containingNamespace)
+        {
+            SymbolKeyData = symbolKeyData;
+            Arity = arity;
+            Name = name;
+            Glyph = glyph;
+            ContainingNamespace = containingNamespace;
+        }
+    }
+
+    internal static partial class ExtensionMethodImportCompletionService
+    {
+        private static readonly char[] s_dotSeparator = new char[] { '.' };
+
+        private static readonly object s_gate = new object();
+        private static Task s_indexingTask = Task.CompletedTask;
+
+        public static async Task<ImmutableArray<SerializableImportCompletionItem>> GetUnimportExtensionMethodsAsync(
+            Document document,
+            int position,
+            ITypeSymbol receiverTypeSymbol,
+            ISet<string> namespaceInScope,
+            bool isExpandedCompletion,
+            CancellationToken cancellationToken)
+        {
+            var ticks = Environment.TickCount;
+            var project = document.Project;
+
+            // This service is only defined for C# and VB, but we'll be a bit paranoid.
+            var client = RemoteSupportedLanguages.IsSupported(project.Language)
+                ? await project.Solution.Workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false)
+                : null;
+
+            var (serializableItems, counter) = client == null
+                ? await GetUnimportExtensionMethodsInCurrentProcessAsync(document, position, receiverTypeSymbol, namespaceInScope, isExpandedCompletion, cancellationToken).ConfigureAwait(false)
+                : await GetUnimportExtensionMethodsInRemoteProcessAsync(client, document, position, receiverTypeSymbol, namespaceInScope, isExpandedCompletion, cancellationToken).ConfigureAwait(false);
+
+            counter.TotalTicks = Environment.TickCount - ticks;
+            counter.TotalExtensionMethodsProvided = serializableItems.Length;
+
+            // TODO: remove this
```

Comment:
>// TODO: remove this [](start = 12, length = 20)

remove?

Reply:
Still needed to to do some perf tests, will remove once done